compiler/rustc_expand/src/proc_macro_server.rs

```diff
@@ -110,10 +110,14 @@ impl FromInternal<(TokenStream, &mut Rustc<'_, '_>)> for Vec<TokenTree<TokenStre
                 tokenstream::TokenTree::Token(token, spacing) => (token, spacing == Joint),
             };
 
+            // Split the operator into one or more `Punct`s, one per character.
+            // The final one inherits the jointness of the original token. Any
+            // before that get `joint = true`.
             let mut op = |s: &str| {
                 assert!(s.is_ascii());
                 trees.extend(s.as_bytes().iter().enumerate().map(|(idx, &ch)| {
```

Comment:
using `bytes` would be a bit shorter and would save a syncatcit deref

Reply:
True, I will change it.

BTW, I think I finally have it clear all the behaviour for operator token joining, i.e. whether to represent things like `>>` as one or two tokens.
- The first lexer (in `rustc_lexer`) keeps operator chars separate, giving split tokens.
- The second lexer (in `rustc_parse`) joins them together.
- The parser splits joined tokens apart when necessary, e.g. when `>>` is two angle brackets for generics rather than a right shift. (In `Parser::break_and_eat`.)
- When converting to the proc-macro layer, joined tokens are resplit. (In `from_internal`.)
- When converting back from the proc-macro layer, split tokens are rejoined. (In `to_internal` and `concat_trees` and `TokenStreamBuilder`.)